pkg/serviceaccounts/oauthclient/oauthclientregistry.go

```diff
@@ -10,27 +12,175 @@ import (
 	"k8s.io/kubernetes/pkg/serviceaccount"
 
 	scopeauthorizer "github.com/openshift/origin/pkg/authorization/authorizer/scope"
+	osclient "github.com/openshift/origin/pkg/client"
 	oauthapi "github.com/openshift/origin/pkg/oauth/api"
 	"github.com/openshift/origin/pkg/oauth/registry/oauthclient"
+	routeapi "github.com/openshift/origin/pkg/route/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/util/sets"
 )
 
 const (
+	// Prefix used for directly specifying redirect URIs for a service account via annotations
 	OAuthRedirectURISecretAnnotationPrefix = "serviceaccounts.openshift.io/oauth-redirecturi."
 	OAuthWantChallengesAnnotationPrefix    = "serviceaccounts.openshift.io/oauth-want-challenges"
+
+	// Prefix used for indirectly specifying redirect URIs using resources for a service account via annotations
+	OAuthRedirectModelAnnotationPrefix              = "serviceaccounts.openshift.io/oauth-redirectmodel."
+	OAuthRedirectModelAnnotationURIPrefix           = OAuthRedirectModelAnnotationPrefix + "uri."
+	OAuthRedirectModelAnnotationResourcePrefix      = OAuthRedirectModelAnnotationPrefix + "resource."
+	OAuthRedirectModelAnnotationURISchemePrefix     = OAuthRedirectModelAnnotationURIPrefix + "scheme."
+	OAuthRedirectModelAnnotationURIPortPrefix       = OAuthRedirectModelAnnotationURIPrefix + "port."
+	OAuthRedirectModelAnnotationURIPathPrefix       = OAuthRedirectModelAnnotationURIPrefix + "path."
+	OAuthRedirectModelAnnotationURIHostPrefix       = OAuthRedirectModelAnnotationURIPrefix + "host."
+	OAuthRedirectModelAnnotationResourceKindPrefix  = OAuthRedirectModelAnnotationResourcePrefix + "kind."
+	OAuthRedirectModelAnnotationResourceNamePrefix  = OAuthRedirectModelAnnotationResourcePrefix + "name."
+	OAuthRedirectModelAnnotationResourceGroupPrefix = OAuthRedirectModelAnnotationResourcePrefix + "group."
+
+	RouteKind = "Route"
+	// TODO add ingress support
+	// IngressKind = "Ingress"
 )
 
+var modelPrefixes = []string{
+	OAuthRedirectModelAnnotationURISchemePrefix,
+	OAuthRedirectModelAnnotationURIPortPrefix,
+	OAuthRedirectModelAnnotationURIPathPrefix,
+	OAuthRedirectModelAnnotationURIHostPrefix,
+	OAuthRedirectModelAnnotationResourceKindPrefix,
+	OAuthRedirectModelAnnotationResourceNamePrefix,
+	OAuthRedirectModelAnnotationResourceGroupPrefix,
+}
+
+// namesToObjMapperFunc is linked to a given GroupKind.
+// Based on the namespace and names provided, it builds a map of resource name to redirect URIs.
+// The redirect URIs represent the default values as specified by the resource.
+// These values can be overridden by user specified data.
+type namesToObjMapperFunc func(namespace string, names sets.String) map[string]redirectURIList
+
+var routeGroupKind = unversioned.GroupKind{Group: routeapi.FutureGroupName, Kind: RouteKind}
```

Comment:
I would start by supporting the current (`""`) group, and later add support for the openshift.io group:
```
var routeGroupKind = routeapi.SchemeGroupVersion.WithKind("Route").GroupKind()
```